### ISSUE TITLE
🐞 - Unable to make a second tour work on another route

### ISSUE THREAD

**Issue by rmoncler-ctl** [hakimio/ngx-ui-tour#103]
### Which ngx-ui-tour-* package is the source of the bug?

ngx-bootstrap

### Description

Hi,

Versions used:
- Angular 9.1
- ngx-ui-tour-core 7.0.2
- ngx-ui-tour-ngx-bootstrap 7.0.2

I have the tour (with `ngx-bootstrap`) working on my main page : it launches automatically on first login, and can be launched manually using a CTA in the header.
So, in order to be able to initialize and start the tour from different components, I created a custom Tour service, injecting the original TourService and implementing most of the methods and observables.

The problem is when I try to add a second tour on another route, it won't work.
Basically, I:
1. build an array of steps
2. pass it to `initialize()`
3. then `start()` the tour

On the second page, the browser's console shows the warning : `Can't attach to unregistered anchor with id tour-secondpage-firststep`. Even though an element has the property `anchorTour="tour-secondpage-firststep"`, properly spelled in both cases.

So I then tried to use the `register()` method from the TourService, resulting in an error poping in the console `Error: anchorId tour-secondpage-firststep already registered!`...
I may have misused this `register()` method because I could not find any documentation or example, especially regarding the TourAnchorDirective parameter.

I also tried using `startAt(stepId)` instead of `start()` : it does not change anything.

Then, I tried setting the isAsync property to true (because while the data is loading, a black transparent overlay with a spinning loader is displayed, so I thought "why not?"). No luck.
Actually, when I do that, there is a slight difference in the behavior :
- The tour seems to work well until it reaches the `start$` observable
- After that, the `stepShow$` observable is never reached
- Even though the steps seem to be properly registered (logging the `initialize$` observable response shows the steps are there)

But that's it, for now I'm stuck with no idea how to make this work a second time at another place in the app, or simply why it does not work.

According to this issue (https://github.com/hakimio/ngx-ui-tour/discussions/58), initializing several tours (each with different steps) is seems to be possible. But is it also true if I'm using the version 7.0.2 ?

Anyway, if you have any idea how I could solve this problem, that would be great!

### Angular version

9.1

### Ngx UI Tour version

7.0.2

### Which browsers have you used?

- [x] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge

### Which operating systems have you used?

- [ ] macOS
- [X] Windows
- [ ] Linux
- [ ] iOS
- [ ] Android

**Comment by hakimio** [comment]
Can you show me your code? How have you implemented your custom tour and how are you using it? Why can't you just use the original tour service in different components? Have you imported `TourModule.forRoot()` in your `AppModule` class?
From the error, I would guess you have multiple instances of the `TourService` and the anchor is not registered in one of the instances. This is not supported. `TourService` should always be global singleton.

**Comment by hakimio** [comment]
The best would be if you created StackBlitz or GitHub reproduction project to demonstrate what you are doing.

**Comment by rmoncler-cotral** [comment]
Thanks a lot for your quick answers, and sorry for the delay on my side.

I tried to start over without a custom service, just to make sure it works when I call the TourService methods directly.
It's likely that it's the way I'm implementing the tour that's causing the error, because I'm getting the same result as before: "`Can't attach to unregistered anchor with id ...`" on the second page. While the first page still works fine.

Building a project or sharing the full code for the example would be difficult, but I will try to explain what I did in detail.

1.
In the AppModule, I added `TourNgxBootstrapModule.forRoot()` to the imports array.
I also added `TourNgxBootstrapModule` (without `.forRoot()`) in other modules that are used by the different components.

2.
In `app.component.html`, I added `<tour-step-template></tour-step-template>`
In the templates, I added the `tourAnchor` attribute on several elements

3.
Then, in the `app.component.ts`, in `ngOnInit()` I subscribe to an observable to launch the tour only when the initial loading is done:
```
ngOnInit() {
  /** ... */

  this.subs.sink = this.loadingService.getLoading().subscribe((loading) => {
    if (!loading) {
      this.launchTour();
    }
  });
}
```

With `launchTour()` defined in the same file, as below (one step only for tour, for testing purpose):

```
public launchTour(): void {
  let viewedTours = [];

  const currentRoute = this.router.url;

  const DASHBOARD_ROUTE = '/dashboard';
  const PERSON_LIST_ROUTE = '/area/person';

  // i18n
  const PREVBTNTITLE = this.i18nService.instant('IN_APP_TUTORIAL.PREV');
  const NEXTBTNTITLE = this.i18nService.instant('IN_APP_TUTORIAL.NEXT');
  const CLOSEBTNTITLE = this.i18nService.instant('IN_APP_TUTORIAL.CLOSE');

  // Declare the tour steps array
  let tourSteps: INgxbStepOption[] = [];

  // Set default values for every step
  const defaultStepValues: INgxbStepOption = {
    prevBtnTitle: PREVBTNTITLE,
    nextBtnTitle: NEXTBTNTITLE,
    endBtnTitle: CLOSEBTNTITLE,
    enableBackdrop: true,
    route: currentRoute,
    isOptional: false,
  };

  if (currentRoute === DASHBOARD_ROUTE && !viewedTours.includes(DASHBOARD_ROUTE)) {
    tourSteps = this.setStepsDashboard(tourSteps);
  } else if (currentRoute === PERSON_LIST_ROUTE && !viewedTours.includes(PERSON_LIST_ROUTE)) {
    tourSteps = this.setStepsPersonList(tourSteps);
  }

  // If there are steps registered, start the tour
  if (tourSteps.length > 0) {
    this.tourService.initialize(tourSteps, defaultStepValues);
    this.tourService.start();
  }
}

/** Set steps to be displayed on the dashboard (first page) */
private setStepsDashboard(tourSteps: INgxbStepOption[]) {
  const tourStepDashboardIntro: INgxbStepOption = {
    stepId: 'tour-dashboard-intro',
    anchorId: 'tour-dashboard-intro',
    content: this.i18nService.instant('IN_APP_TUTORIAL.DASHBOARD.INTRO.CONTENT'),
    title: this.i18nService.instant('IN_APP_TUTORIAL.DASHBOARD.INTRO.TITLE'),
  };
  
  tourSteps.push(tourStepDashboardIntro);

  return tourSteps;
}

/** Set steps to be displayed on the person list (second page) */
private setStepsPersonList(tourSteps: INgxbStepOption[]) {
  const tourStepPersonListIntro: INgxbStepOption = {
    stepId: 'tour-personlist-intro',
    anchorId: 'tour-personlist-intro',
    content: this.i18nService.instant('IN_APP_TUTORIAL.PERSONLIST.INTRO.CONTENT'),
    title: this.i18nService.instant('IN_APP_TUTORIAL.PERSONLIST.INTRO.TITLE'),
  };

  tourSteps.push(tourStepPersonListIntro);

  return tourSteps;
}
```

And with that, I still got the same error as before, so I probably made a mistake during the implementation but I can't find what it is.

(I removed a few `console.log` for a better readability, but the `tourSteps` array passed to `initialize()` is correctly defined in both cases)

**Comment by rmoncler-cotral** [comment]
Also, I forgot and I'm not sure if it would help, but on the first page (working), when I inspect the DOM I see;
```
<h2 _ngcontent-fdu-c250="" touranchor="tour-dashboard-intro" ng-reflect-tour-anchor="tour-dashboard-intro">Dashboard</h2>
```

And on the second page, the HTML element has no "ng-reflect-..." attribute:
```
<h2 _ngcontent-fdu-c218="" touranchor="tour-personlist-intro" >List of people</h2>
```

I'm not sure why, or if it could be related to my problem.

**Comment by hakimio** [comment]
If you use [`anchorRegister$` observable](https://hakimio.github.io/ngx-ui-tour/ngx-bootstrap/API#events), do you see that your "tour-personlist-intro" is registered successfully?

**Comment by rmoncler-ctl** [comment]
I have imported `TourNgxBootstrapModule` in the module where the second page is declared.

It seems the `anchorRegister$` observable is never reached. On the first page, I do se every step being registered, but nothing happens in this observable on the second page.

`delayAfterNavigation` seems to have no effect.

**Comment by hakimio** [comment]
> It seems the anchorRegister$ observable is never reached. On the first page, I do se every step being registered, but nothing happens in this observable on the second page.

What do you mean? You should only subscribe to the observable once and then see the anchors getting registered when they appear on the page.

Where are you importing `TourService` from? `ngx-ui-tour-core` or `ngx-ui-tour-ngx-bootstrap`? Do you have any imports from `ngx-ui-tour-core`?

**Comment by rmoncler-ctl** [comment]
More precisely, I added the following code immediatly below `this.tourService.start()`:
```
this.subs.sink = this.tourService.anchorRegister$.subscribe((res) => {
  console.log('anchorRegister$');
  console.log(res);
});
```

On the first page, I have:
![image](https://user-images.githubusercontent.com/95758743/216071108-01b3123d-6f79-44a0-bdba-1930a16a10fb.png)

But on the second page, nothing shows up, no "anchorRegister$" logged, only the warning "`Can't attach to unregistered anchor...`"

**Comment by hakimio** [comment]
So, are you importing the `TourService` from the correct package?

**Comment by rmoncler-ctl** [comment]
Yes sorry, I import only from `ngx-ui-tour-ngx-bootstrap`, nothing from `ngx-ui-tour-core`.

**Comment by hakimio** [comment]
- Try to initialize the `TourService` with all the steps once in your `AppComponent` and then only call `start()` in your components. Add a single subscription to `anchorRegister$` in `AppComponent`, to check that all the anchors are registered correctly.
- If you want me to take a look at your current code, create `StackBlitz` or `GitHub` project reproducing the issue.

**Comment by rmoncler-ctl** [comment]
I found it: the routing (and its modules) of our application are a bit specific (not to say a bit twisted), and that's where the problem was.
I tried importing and exporting the `TourNgxBootstrapModule` in the `app-routing.module.ts` file, and that seems to have solved the problem.

Again, thank you very much for your time and your help. At least it made me double(triple)-check every step of my implementation.
